container/rules/debian_pkg_tar.bzl

```diff
@@ -0,0 +1,188 @@
+      toolchain_container rule
+      keys: keys aggregated from each language_tool_layer by toolchain_container rule
+      download_pkgs_output_tar: output tar file generated by download_pkgs rule to
+      override default output_tar name
+      download_pkgs_output_script: output script generated by download_pkgs rule to
+      override default output_script name
+    """
+
+    # Prepare base image for the download_pkgs rule.
+    download_base = ctx.files.base[0]
+
+    # Create an intermediate image with additional gpg keys used to download packages.
+    if keys != []:
+        image_with_keys = "%s_with_keys" % ctx.attr.name
+
+        # Declare intermediate output file generated by add_apt_key rule.
+        image_with_keys_output_executable = ctx.actions.declare_file(image_with_keys)
+        image_with_keys_output_tarball = ctx.actions.declare_file(image_with_keys + ".tar")
+        image_with_keys_output_layer = ctx.actions.declare_file(image_with_keys + "-layer.tar")
+
+        _apt_key.implementation(
+            ctx,
+            name = image_with_keys,
+            image_tar = ctx.files.base[0],
+            keys = keys,
+            output_executable = image_with_keys_output_executable,
+            output_tarball = image_with_keys_output_tarball,
+            output_layer = image_with_keys_output_layer,
+        )
+        download_base = image_with_keys_output_tarball
+
+    # Declare intermediate output file generated by download_pkgs rule.
+    output_executable = ctx.actions.declare_file(ctx.attr.name + "-output_executable.sh")
+    download_pkgs_output_tar = download_pkgs_output_tar or ctx.attr.name + ".tar"
+    download_pkgs_output_script = download_pkgs_output_script or ctx.attr.name + ".sh"
+    output_tar = ctx.actions.declare_file(download_pkgs_output_tar)
+    output_script = ctx.actions.declare_file(download_pkgs_output_script)
+
+    # download_pkgs rule consumes 'packages' and 'additional_repos'.
+    _download_deb_pkgs.implementation(
+        ctx,
+        image_tar = download_base,
+        packages = packages,
+        additional_repos = additional_repos,
+        output_executable = output_executable,
+        output_tar = output_tar,
+        output_script = output_script,
+    )
+
+    return struct(
+        providers = [
+            InstallableTarInfo(
+                installables_tar = output_tar,
+            ),
+        ],
+    )
+
+def _aggregate_debian_pkgs_impl(ctx):
+    """Implementation for the aggregate_debian_pkgs rule.
+
+    aggregate_debian_pkgs rule produces a tarball with all debian packages declared
+    in the language_tool_layer(s) this rule depends on.
+
+    Args:
+      ctx: ctx as the same as for container_image + list of language_tool_layer(s)
+      https://github.com/bazelbuild/rules_docker#container_image
+    """
+
+    # If the rule is used directly, aggregate packages, additional_repos, keys
```

Comment:
why condition this on "If the rule is used directly"

Reply:
This was a comment of a previous commit, will remove it here, cause this rule is only used for generating a tarball.
Thanks for pointing out.